### PR TITLE
adds an option to provide raw bson output

### PR DESCRIPTION
**Pull request by lrlna**
This allows for both `collection.find` and `collection.aggregate` to accept `raw` option and give us back raw bson to work with. `raw` defaults to false.

### CODE REVIEW

File: test/index.test.js [hunk 1]
```diff
@@ -148,6 +148,93 @@ describe('mongodb-collection-sample', function() {
...
+
+    after(function(done) {
+      if (!db) {
+        return done();
+      }
+      db.dropCollection('haystack', done);
+    });
+
+    it('buffer does not overflow', function(done) {
+      sample(db, 'haystack', { size: 1000000, raw: true })
```
Comment (lrlna) [inline test/index.test.js]: i just decided to check if we can still sample with 1mil. seems a bit naive. is there anything specific i can check to see if buffer is overflowing?

Comment (imlucas) [reply test/index.test.js]: > 🚧 stream of conscious-too-specific answer :) 

hmm... can't think of any other than [hunting down where memory leak](https://marmelab.com/blog/2018/04/03/how-to-track-and-fix-memory-leak-with-nodejs.html) is coming from 😬 

Something that comes to mind: the actual sampling math wants `sampleSize` much lower (`5` I think $sample default i spec'd was 5)  The way the reservoir sampler is implemented here would be a big chunk of work to get rid of all the leaks. 

Even though its not a sample, the `give me 1mm documents in a bson stream without having to use something else` is neat actually. 

You could check if `opts.sampleSize > 999999` or some other flag/value on the top-level  ➡️ 
If yes, skip the server version check and return a new `BigSampler()` class that impls `BaseSampler` ➡️ 
return `db.aggregate([{$match: this.query}, {$project... etc.])` cursor wrapped in a readable stream.

File: test/index.test.js [hunk 2]
```diff
@@ -148,6 +148,93 @@ describe('mongodb-collection-sample', function() {
...
+          done();
+        }));
+    });
+  });
+
+  // This test creates 2mil documents and samples 1mil of those to make sure
+  // buffer doesn't overflow. Only creating 2mil documents, since any more
+  // causes v8 to run out memory in heap.
+  // FYI: takes a long time to run.
+  describe('raw buffer over a large set of documents', function() {
```
Comment (imlucas) [inline test/index.test.js]: one thing that might be work a try: wrap this whole block in a `if (process.env.TEST_BIG_SAMPLER)` or something so its only run if you tell it to specifically and add details to `README`

Comment (lrlna) [reply test/index.test.js]: ohhhhhh yes good idea!

File: test/index.test.js (hunk 1 above)
Comment (lrlna) [reply test/index.test.js]: i think i lost you here 😅😅. I don't particular have a memory leak with this particular test (I don't think). And the test also running this in the `Native Sampler`. 

But you're saying I should be checking straight up if the `BaseSampler` is doing its job?

**Comment by lrlna** [comment]
What i get with a raw flag is a `cursor`,`firstBatch` and then the documents. 
in js: (gotten from: `buf.toString()`)
Screen Shot 2019-03-17 at 22 53 31
in rust: (gotten from: `Uint8Array` -> `&[u8]` -> `bson::decode_document(&mut [u8])`)
Screen Shot 2019-03-17 at 23 01 38



What I thought I would be getting is same thing as with `raw: false`, aka data dump of however many documents but just as a bson buffer. Does that seem like a correct assumption?

**Comment by lrlna** [comment]
@imlucas i made some changes to this including:

1. adding `process.env` check to the big sampling test
2. i am choosing to run the native-sampler if we are asking for raw documents
3. there is also now a `pipe` to convert the raw cursor we get into individual raw documents.

**Comment by imlucas** [comment]
Published in `4.4.0`

**Pull request merged**